content/posts/faster-rfft-python/index.md

```diff
@@ -0,0 +1,453 @@
+
+### Fourth Attempt: Reuse the RFFT object (v4)
+As our last attempt, we can try to reuse the RFFT object across multiple calls when the input size does not change. This avoids the overhead of creating a new RFFT object on every call.
+
+```python
+class rfft_caller_v4:
+    def __init__(self):
+        self.real_arr = None
+        self.complex_arr = None
+        self.rfft_obj = None
+    
+    
+    def __call__(self, T):
+        if self.real_arr is None or len(T) != len(self.real_arr):
+            self.real_arr = pyfftw.empty_aligned(len(T), dtype='float64')    
+            self.complex_arr = pyfftw.empty_aligned(len(T) // 2 + 1, dtype='complex128')
+
+            self.rfft_obj = pyfftw.FFTW(
+                    self.real_arr,
+                    self.complex_arr,
+                    direction='FFTW_FORWARD',
+                    flags=('FFTW_MEASURE',),
+                    threads=1,
+                ) 
+           
+        self.real_arr[:] = T
+        self.rfft_obj.execute()
+
+        return self.complex_arr
+
+_rfft_caller_v4 = rfft_caller_v4()
+```
+
+And we can check the performance by updating the dictionary `rfft_callers`, and running the plot script.
+
+```python
+rfft_callers = {
+    'v0': rfft_caller_v0,
+    'v1': rfft_caller_v1,
+    'v2': rfft_caller_v2,
+    'v3': rfft_caller_v3,
+    'v4': rfft_caller_v4,
+}
+```
+
+![Performance Gain](Figure_v4.png)
+
+As shown in the performance plot above, reusing the RFFT object resulted in a further performance improvement. For arrays with input size `<= 2^8`, the current version is 30-35x faster than the baseline! From that point onward, the performance gain reduces as the input size increases, but it is still around 2x faster for input size `2^24`. This is a significant improvement compared to our baseline implementation! If the application requires frequent RFFT computations on arrays of same size, then this approach can be very beneficial as one can store the `rfft_caller_v4` instance and reuse it across multiple calls.
+
+
+## Conclusion
+In this post, we explored different ways to optimize the performance of RFFT computation using pyFFTW in Python. We started with a baseline implementation and gradually improved it by applying various optimizations, such as byte-aligning the input array, using the `pyfftw.FFTW` object directly, and reusing the RFFT object across multiple executions. We did not explore multithreading here, but it may offer further performance gains, especially for large arrays.
+
+
+## BONUS: Anything else?
+During my exploration, I also tried the following things based on what I read in the [FFTW documentation](https://www.fftw.org/fftw3_doc/Planner-Flags.html). 
+
+* v5: passing the `FFTW_DESTROY_INPUT`flag: According to the documentation, this flag indicates that the input array can be destroyed during the computation. This allows FFTW to use the memory of the input array if needed, which might lead to performance improvements.
```

Comment:
Add the following part that is mentioned in the document: 

> The default behaviour is, if possible, to preserve the input. In the case of the 1D Backwards Real transform, this may result in a performance hit. 

So, in "Forward" transform, we should expect to see no improvement. 

----

We can add that here, or in the part where we explain the performance plot.